quaint/src/connector/postgres/native/cache.rs

```diff
@@ -0,0 +1,394 @@
+/// cannot be re-used when tracing information is present.
+#[derive(Debug)]
+pub struct TracingLruCache {
+    cache: InnerLruCache<Arc<QueryMetadata>>,
+}
+
+impl TracingLruCache {
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cache: InnerLruCache::with_capacity(capacity),
+        }
+    }
+}
+
+#[async_trait]
+impl QueryCache for TracingLruCache {
+    type Query<'a> = TypedQuery<'a>;
+
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<TypedQuery<'a>, Error> {
+        let sql_without_traceparent = strip_query_traceparent(sql);
+
+        let metadata = match self.cache.get(sql_without_traceparent, types).await {
+            Some(metadata) => metadata,
+            None => {
+                let stmt = client.prepare_typed(sql_without_traceparent, types).await?;
+                let metdata = Arc::new(QueryMetadata::from(&stmt));
+                self.cache.insert(sql_without_traceparent, types, metdata.clone()).await;
+                metdata
+            }
+        };
+        Ok(TypedQuery::from_sql_and_metadata(sql, metadata))
+    }
+
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error> {
+        client.prepare_typed(sql, types).await
+    }
+}
+
+impl From<CacheSettings> for TracingLruCache {
+    fn from(settings: CacheSettings) -> Self {
+        Self::with_capacity(settings.capacity)
+    }
+}
+
+/// Settings related to query caching.
+#[derive(Debug)]
+pub struct CacheSettings {
+    pub capacity: usize,
+}
+
+/// Key uniquely representing an SQL statement in the prepared statements cache.
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct QueryKey {
+    /// Hash of a string with SQL query.
+    sql: u64,
+    /// Combined hash of types for all parameters from the query.
+    types_hash: u64,
+}
+
+impl QueryKey {
+    fn new<S: BuildHasher>(st: &S, sql: &str, params: &[Type]) -> Self {
+        Self {
+            sql: st.hash_one(sql),
+            types_hash: st.hash_one(params),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct InnerLruCache<V> {
+    cache: Mutex<LruCache<QueryKey, V>>,
+    state: RandomState,
+}
+
+impl<V> InnerLruCache<V> {
+    fn with_capacity(capacity: usize) -> Self {
+        Self {
+            cache: Mutex::new(LruCache::new(capacity)),
+            state: RandomState::new(),
+        }
+    }
+
+    async fn get(&self, sql: &str, types: &[Type]) -> Option<V>
+    where
+        V: Clone,
+    {
+        let mut cache = self.cache.lock().await;
+        let capacity = cache.capacity();
+        let stored = cache.len();
+
+        let key = QueryKey::new(&self.state, sql, types);
+        // we call `get_mut` because LRU requires mutable access for lookups
+        match cache.get_mut(&key) {
+            Some(value) => {
+                tracing::trace!(
+                    message = "CACHE HIT!",
```

Comment:
fwiw these two logs in caps have always annoyed me when I saw them in code 😆

Reply:
I can change the message to something less annoying if you have any suggestions 😅

Reply:
I think just `message = "cache hit"` and `message = "cache miss"` is fine but I don't really care that much, your call 😄

Reply:
ok, updated it